### DISCUSSION TITLE
How to apply context on of top of zustand (redux way)

### DISCUSSION THREAD

**Discussion by igortas** [pmndrs/zustand#2111] [Q&A]
Hi,
I'm using zustand in more old fashioned redux way, so I have issues to wrap the store with context provider.

If I have something like this how I can wrap it with context provider and use that one:

```
function AlertReducer(state: AlertState, action: AlertAction) {
	switch (action.type) {
		case 'ALERT/SET_SUCCESS': {
			const { successMessage, autoHideMs } = action.payload;
             ........
}
```

```export const useAlertStore = create(redux(AlertReducer, initialAlertState))```

**Comment by dai-shi** [comment]
Try using `createStore` and `useStore`:
```jsx
import { createContext, useContext } from 'react'
import { createStore, useStore } from 'zustand'

const alertStore = createStore(redux(AlertReducer, initialAlertState))
const AlertStoreContext = createContext(null)
export const Provider = ({ children }) => <AlertStoreContext.Provider value={alertStore}>{children}</AlertStoreContext.Provider>
export const useSelector = (selector) => useStore(useContext(AlertStoreContext), selector)
```

**Comment by igortas** [comment]
@dai-shi 
Hi once again

How to use this example with ```typescript``` properly?

Because ```useStore``` is deprecated, I'm using ```useStoreWithEqualityFn``` now, but I'm having issues with ```unknown```.

```
type TabsContextState = TabsState & { dispatch: (selector: TabAction) => TabAction };

const TabsContext = createContext<TabsContextState | null>(null);

export const TabsProvider = ({ children }: { children: ReactNode }) => {
  const store = create(redux(tabsReducer, initialTabsState));
  return (
    <TabsContext.Provider value={store.getState() as TabsContextState}>
      {children}
    </TabsContext.Provider>
  )
}

type Selector = (selector: TabAction) => TabAction;

export const useTabsStore = (selector: Selector) => {
  const store = useContext(TabsContext) as TabsContextState;
  return useStoreWithEqualityFn(store, selector)
}
```

```TS``` warning:

```
Argument of type 'TabsContextState' is not assignable to parameter of type 'WithReact<StoreApi<unknown>>'.
  Type 'TabsContextState' is missing the following properties from type 'StoreApi<unknown>': setState, getState, subscribe, destroy
```

**Reply by dai-shi** [reply]
`useStore` isn't deprecated.

Can you create a reproduction with modifying this? https://tsplay.dev/WGYxvN

**Reply by igortas** [reply]
@dai-shi My question was more towards how and where to create the zustand store when I'm wrapping the components with context provider...
```const store = create(redux(tabsReducer, initialTabsState));```
Is this line fine and should be just above the ```TabsContext.Provider```, because when I'm doing like that and than I have some handlers like, ```onClick```, nothing is rerendered... I see that ```dispatch``` is working fine, but ```React``` not rerendering...

**Reply by dai-shi** [reply]
I thought it's TS typing question.

No, it's not fine if the Provider component re-renders.
useRef is one solution:
https://tsplay.dev/weyDEw
```ts
export const TabsProvider = ({ children }: { children: ReactNode }) => {
  const storeRef = useRef();
  if (!storeRef.current) {
    storeRef.current = createStore(redux(tabsReducer, initialTabsState));
  }
  return <TabsContext.Provider value={storeRef.current}>{children}</TabsContext.Provider>;
};
```

**Comment by igortas** [comment]
@dai-shi I'll do it later. Is this stable code for context api on top of zustand? I have issues with rerender when updating new values... no new values context provider sends to consumer components...

**Reply by dai-shi** [reply]
Yes, using context api with zustand store is recommended for non-global use case and it's stable.
